src/python/pants/backend/docker/target_types.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.engine.target import COMMON_TARGET_FIELDS, Dependencies, Sources, StringField, Target
+
+
+class DockerImageSources(Sources):
+    default = ("Dockerfile",)
+    expected_num_files = 1
+    help = "Name of the Dockerfile to use when building the Docker image."
```

Comment:
This isn't quite true, is it? It's not just the name, it's the actual Dockerfile that you're going to be building. Maybe something more like this?

> The Dockerfile to build.
>
> This must be exactly one file, usually named `Dockerfile`.

I can envision us in the future documenting restrictions here like "The Dockerfile must not do x".

Reply:
Hmm.. you are not building the Dockerfile. You are using the Dockerfile as "blueprint" for the image you build. 

Not sure how this maps to potential future restrictions on what not to do?

Reply:
Also, I imagine that we can allow more than one file, in which case we'd simply concatenate them into one, resulting in a multistage build. But that is a future feature.. ;)

Reply:
> you are not building the Dockerfile. You are using the Dockerfile as "blueprint" for the image you build.

Hm, okay. Although that is still more than suggested by "Name of the Dockerfile". Is there another way to describe what this does, perhaps?

> Not sure how this maps to potential future restrictions on what not to do?

Sorry, I was thinking out loud. You can ignore that part :)

Reply:
Well, this is how Docker describes the `build` action:

> The docker build command builds Docker images from a Dockerfile and a “context”.

And the `--file` option could be the source for the leading part of the above help text..

> --file , -f |   | Name of the Dockerfile (Default is 'PATH/Dockerfile')

So I don't think we have to overdo Docker in this regard, and just say that this parameter names the `Dockerfile` to use.

https://docs.docker.com/engine/reference/commandline/build/

Reply:
But my point being that you are actually going to consume this source file, right? Like feed it directly to Docker? 

An alternative design for Docker is to have users express everything directly in a BUILD file, and then Pants creates a synthetic Dockerfile for you. When I see "Name of the Dockerfile", it's making me think of that model, that this is doing nothing more than setting the name of the artifact, when really the file is going to be fed directly into Docker.

Reply:
Yes, it has to exist, for now. But we could potentially (and I think there is good use for it) have all the required fields in the `docker_image` target to be able to generate the `Dockerfile`, but then we wouldn't even need to ask for the name of the file, could just settle for naming it `Dockerfile`.

But as this is for the `sources` field, isn't it implied the file has to exist?

```
docker_image(
    sources=("MyCustomDockerfileName",)
)
```

I would quickly learn if this file was missing, too...

Reply:
Ooohhh... I think we look at it differently. What I see is "The name of the Dockerfile to use", whereas it feels like you see it as "The name to use for the Dockerfile". Could that be it?

Reply:
How about this?

> "The Dockerfile to use when building the Docker image."

Reply:
Ah yes, I think that is it! I'm not sure why I'm reading it that way..hm. Sorry about the confusion!

As silly as this is, adding "the" to the beginning makes me less confused I think? 

Thanks for bearing with me on this one

Reply:
It's closer to what you suggested, but still not implying that we're building the Dockerfile. ;)

Reply:
Sounds good! I like that wording. Thank you

Reply:
> Could that be it?
Re-reading your previous comment confirms that.. I went ahead and tweaked the help text slightly in the direction of your suggestion. See if it lands OK.. :)

Reply:
Nvm, your replies didn't show up here for some reason...

Reply:
One should not forget, that if you get confused, so could the users.

Reply:
How about `The Dockerfile to use when building the Docker image.` ? Because theoretically it could be in a subdir, so "name" isn't completely accurate.

Reply:
That's exactly what we landed in :)